velox/common/memory/ArbitrationParticipant.h

```diff
@@ -0,0 +1,348 @@
+  // arbitration participant for access.
+  std::optional<ScopedArbitrationParticipant> lock();
+
+  /// Returns the corresponding query memory pool.
+  MemoryPool* pool() const {
+    return pool_;
+  }
+
+  /// Returns the current capacity of the query memory pool.
+  uint64_t capacity() const {
+    return pool_->capacity();
+  }
+
+  /// Gets the capacity growth targets for 'requestBytes' growth request. The
+  /// capacity growth target is set to have a coarser granularity. It can help
+  /// to reduce the number of future grow calls, and hence reducing the number
+  /// of unnecessary memory arbitration requests.
+  void getGrowTargets(
+      uint64_t requestBytes,
+      uint64_t& maxGrowBytes,
+      uint64_t& minGrowBytes) const;
+
+  /// Returns the unused free memory capacity that can be reclaimed from the
+  /// query memory pool by shrink.
+  uint64_t reclaimableFreeCapacity() const;
+
+  /// Returns the used memory capacity that can be reclaimed from the query
+  /// memory pool through disk spilling.
+  uint64_t reclaimableUsedCapacity() const;
+
+  /// Checks if the query memory pool can grow 'requestBytes' from its current
+  /// capacity under the max capacity limit.
+  bool checkCapacityGrowth(uint64_t requestBytes) const;
+
+  /// Invoked to grow the query memory pool capacity by 'growBytes' and commit
+  /// used reservation by 'reservationBytes'. The function throws if the growth
+  /// fails.
+  bool grow(uint64_t growBytes, uint64_t reservationBytes);
+
+  /// Invoked to release the unused memory capacity by reducing its capacity. If
+  /// 'reclaimAll' is true, the function releases all the unused memory capacity
+  /// from the query memory pool without regarding to the minimum free capacity
+  /// restriction.
+  uint64_t shrink(bool reclaimAll = false);
+
+  // Invoked to reclaim used memory from this memory pool with specified
+  // 'targetBytes'. The function returns the actually freed capacity.
+  uint64_t reclaim(uint64_t targetBytes, uint64_t maxWaitTimeMs) noexcept;
+
+  /// Invoked to abort the query memory pool and returns the reclaimed bytes
+  /// after abort.
+  uint64_t abort(const std::exception_ptr& error) noexcept;
+
+  /// Returns true if the query memory pool has been aborted.
+  bool aborted() const {
+    std::lock_guard<std::mutex> l(stateLock_);
+    return aborted_;
+  }
+
+  /// Invoked to wait for the pending memory reclaim or abort operation to
+  /// complete within a 'maxWaitTimeMs' time window. The function returns false
+  /// if the wait has timed out.
+  bool waitForReclaimOrAbort(uint64_t maxWaitTimeMs) const;
+
+  /// Invoked to start arbitration operation 'op'. The operation needs to wait
+  /// for the prior arbitration operations to finish first before executing to
+  /// ensure the serialized execution of arbitration operations from the same
+  /// query memory pool.
+  void startArbitration(ArbitrationOperation* op);
```

Comment:
Shall we rename this method to enqueueArbitration() as technically the arbitration is not started by this method?

Reply:
When the function returns, the operation is started so I think it is better to call this startArbitration.